### PR TITLE
Implementation of PS-3969 - 5.6

### PR DESCRIPTION
**Pull request by altmannmarcelo**
This commit adds a variable to track when an audit log entry was either
 dropped or written directly to the file due to its size being bigger
than audit_log_buffer_size

### CODE REVIEW

File: mysql-test/t/audit_log_PS-3969.test [hunk 1]
```diff
@@ -0,0 +1,21 @@
+# Bug PS-3969: Adding a status variable to catch when statement was too big to fit buffer.
```
Comment (laurynas-biveinis) [inline mysql-test/t/audit_log_PS-3969.test]:
```suggestion
--source include/not_embedded.inc
# Bug PS-3969: Adding a status variable to catch when statement was too big to fit buffer.
```
server restarts are not supported with embedded server

File: mysql-test/t/audit_log_PS-3969.test [hunk 2]
```diff
@@ -0,0 +1,21 @@
...
...YyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonj";
```
Comment (laurynas-biveinis) [inline mysql-test/t/audit_log_PS-3969.test]:
```suggestion
--disable_query_log
--disable_result_log
SELECT "4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonjTjv4IHYyOZMEAUJcURiOO9nLoceKvbRi3HSZZObmonj";
--enable_result_log
--enable_query_log
```
At least the .result file will avoid embedding this large string

File: plugin/audit_log/audit_log.c [hunk 3]
```diff
@@ -1746,7 +1748,9 @@ static struct st_mysql_audit audit_log_descriptor=
 
 static struct st_mysql_show_var audit_log_status_variables[]=
 {
-  { 0, 0, 0}
+  {"Audit_buffer_size_too_small",
+   (char*) &audit_buffer_size_too_small,
+   SHOW_LONG}
```
Comment (laurynas-biveinis) [inline plugin/audit_log/audit_log.c]:
```suggestion
   SHOW_LONG},
   { 0, 0, 0}
```
Keep the terminating NULL entry

File: plugin/audit_log/buffer.c [hunk 4]
```diff
@@ -189,6 +189,7 @@ int audit_log_buffer_write(audit_log_buffer_t *log, const char *buf, size_t len)
       log->write_func(log->write_func_data, buf, len, LOG_RECORD_COMPLETE);
       audit_log_buffer_resume(log);
     }
+    audit_buffer_size_too_small++;
```
Comment (laurynas-biveinis) [inline plugin/audit_log/buffer.c]: Must be accessed in a thread safe way: for 5.6 and 5.7, look up my_atomic.h, for 8.0, std::atomic

File: mysql-test/t/audit_log_PS-3969.test [hunk 5]
```diff
@@ -0,0 +1,21 @@
...
```
Comment (laurynas-biveinis) [inline mysql-test/t/audit_log_PS-3969.test]: Likewise

File: plugin/audit_log/audit_log.c [hunk 6]
```diff
@@ -1746,7 +1748,9 @@ static struct st_mysql_audit audit_log_descriptor=
 
 static struct st_mysql_show_var audit_log_status_variables[]=
 {
-  { 0, 0, 0}
+  {"Audit_buffer_size_too_small",
```
Comment (laurynas-biveinis) [inline plugin/audit_log/audit_log.c]: Should start with plugin name prefix (which is "audit_log")
"too_small" does not make it obvious that this is a counter, please consider other names

File: mysql-test/t/audit_log_PS-3969.test (hunk 1 above)
Comment (laurynas-biveinis) [inline mysql-test/t/audit_log_PS-3969.test]: Rename to audit_log_<status_var_name>.test

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

**Comment by laurynas-biveinis** [comment]
Have you run the rest of the MTR suite?

**Comment by altmannmarcelo** [comment]
Ready for the next review - https://ps.cd.percona.com/job/percona-server-5.6-param/51/

File: plugin/audit_log/audit_log.h [hunk 7]
```diff
@@ -19,6 +19,8 @@
 
 #include <m_ctype.h>
 
+extern volatile int64 audit_log_buffer_size_overflow;
```
Comment (laurynas-biveinis) [inline plugin/audit_log/audit_log.h]:
```suggestion
extern uint64 audit_log_buffer_size_overflow;
```

File: plugin/audit_log/audit_log.c [hunk 8]
```diff
@@ -68,6 +68,8 @@ static char *audit_log_include_accounts= NULL;
 static char *audit_log_exclude_commands= NULL;
 static char *audit_log_include_commands= NULL;
 
+volatile int64 audit_log_buffer_size_overflow = 0;
```
Comment (laurynas-biveinis) [inline plugin/audit_log/audit_log.c]: Atomic vars have little to do with volatile; SHOW_LONG is actually unsigned, which is OK for us:
```suggestion
uint64 audit_log_buffer_size_overflow = 0;
```

File: plugin/audit_log/buffer.c [hunk 9]
```diff
@@ -118,7 +118,7 @@ void *audit_log_flush_worker(void *arg)
 audit_log_buffer_t *audit_log_buffer_init(size_t size, int drop_if_full,
                                  audit_log_write_func write_func, void *data)
 {
-  audit_log_buffer_t *log= (audit_log_buffer_t*) 
+  audit_log_buffer_t *log= (audit_log_buffer_t*)
```
Comment (laurynas-biveinis) [inline plugin/audit_log/buffer.c]: Spurious diff. Or is it a result of clang-format on the file?

File: plugin/audit_log/buffer.c [hunk 10]
```diff
@@ -189,6 +189,7 @@ int audit_log_buffer_write(audit_log_buffer_t *log, const char *buf, size_t len)
       log->write_func(log->write_func_data, buf, len, LOG_RECORD_COMPLETE);
       audit_log_buffer_resume(log);
     }
+    my_atomic_add64(&audit_log_buffer_size_overflow, (int64)1);
```
Comment (laurynas-biveinis) [inline plugin/audit_log/buffer.c:193]: Redundant cast:

```suggestion
    my_atomic_add64(&audit_log_buffer_size_overflow, 1);
```

File: plugin/audit_log/buffer.h [hunk 11]
```diff
@@ -20,6 +20,7 @@
 
 #include <string.h> // for size_t
 #include "logger.h"
+#include <my_atomic.h>
```
Comment (laurynas-biveinis) [inline plugin/audit_log/buffer.h]: Please include where it's used instead

**Review by laurynas-biveinis — CHANGES_REQUESTED** [review]

File: plugin/audit_log/buffer.c (hunk 10 above)
Comment (altmannmarcelo) [reply plugin/audit_log/buffer.c:193]: > Redundant cast:
> 
> Suggested change Beta 
> [Give us feedback](https://www.research.net/r/SuggestedChanges)
>     my_atomic_add64(&audit_log_buffer_size_overflow, (int64)1);
>     my_atomic_add64(&audit_log_buffer_size_overflow, 1);

@laurynas-biveinis if I don't cast, it gives me an error:

```
/home/marcelo/git/percona-server/plugin/audit_log/buffer.c: In function ‘audit_log_buffer_write’:
/home/marcelo/git/percona-server/plugin/audit_log/buffer.c:193:54: error: incompatible type for argument 2 of ‘my_atomic_add64’
     my_atomic_add64(&audit_log_buffer_size_overflow, 1);
                                                      ^
In file included from /home/marcelo/git/percona-server/plugin/audit_log/buffer.c:22:0:
/home/marcelo/git/percona-server/include/my_atomic.h:168:24: note: expected ‘U_64 {aka union <anonymous>}’ but argument is of type ‘int’
 static inline int ## S my_atomic_add ## S(                      \
                        ^
/home/marcelo/git/percona-server/include/my_atomic.h:211:1: note: in expansion of macro ‘make_atomic_add’
 make_atomic_add(64)
 ^~~~~~~~~~~~~~~
``` 

For 5.7 it works fine, only 5.6 that requires the implicit casting.

**Comment by altmannmarcelo** [comment]
ready for next review - https://ps.cd.percona.com/job/percona-server-5.6-param/58/

**Review by laurynas-biveinis — APPROVED** [review]

**Comment by altmannmarcelo** [comment]
replaced by #3056

**Pull request closed without merging**